db/hints/sync_point_service.cc

```diff
@@ -0,0 +1,442 @@
+                    task.schedule_expiry(completed_coordinator_task_ttl);
+                });
+            });
+        });
+
+        return make_ready_future<sync_point_id>(id);
+    });
+}
+
+future<> sync_point_service::wait_for_sync_point(sync_point_id id, clock_type::time_point timeout) {
+    return container().invoke_on(0, [id, timeout] (sync_point_service& sps) {
+        sps.check_if_is_running();
+
+        auto it = sps._coordinator_tasks.find(id);
+        if (it == sps._coordinator_tasks.end()) {
+            throw std::runtime_error(format("Coordinator hint sync point with id {} does not exist", id));
+        }
+        return it->second.wait_for_result(timeout);
+    });
+}
+
+future<> sync_point_service::delete_sync_point(sync_point_id id) {
+    return container().invoke_on(0, [id] (sync_point_service& sps) {
+        sps.check_if_is_running();
+
+        auto it = sps._coordinator_tasks.find(id);
+        if (it == sps._coordinator_tasks.end()) {
+            throw std::runtime_error(format("Coordinator hint sync point with id {} does not exist", id));
+        }
+        return it->second.cancel_and_finish();
+    });
+}
+
+std::vector<sync_point_id> sync_point_service::list_sync_points() const {
+    std::vector<sync_point_id> ids;
+    ids.reserve(_coordinator_tasks.size());
+    for (auto& p : _coordinator_tasks) {
+        ids.push_back(p.first);
+    }
+    return ids;
+}
+
+void sync_point_service::register_rpc_verbs() {
+    auto& ms = _messaging;
+
+    ms.register_hint_sync_point_create([this] (sync_point_create_request request) -> future<sync_point_create_response> {
+        return container().invoke_on(0, [request = std::move(request)] (sync_point_service& sps) mutable {
+            sps.check_if_is_running();
+
+            const auto sync_point_id = request.sync_point_id;
+            auto target_endpoints = std::move(request.target_endpoints);
+
+            if (sps._follower_tasks.size() >= max_follower_sync_points) {
+                throw std::runtime_error(format("Too many follower sync points; the limit is {}", max_coordinator_sync_points));
+            }
+            if (sps._follower_tasks.contains(sync_point_id)) {
+                throw std::runtime_error(format("Follower hint sync point with id {} already exists", sync_point_id));
+            }
+
+            auto [it, emplaced] = sps._follower_tasks.try_emplace(sync_point_id, sps._follower_tasks, "follower", sync_point_id);
+            assert(emplaced);
+            auto& task = it->second;
+            task.run([&sps, sync_point_id, &target_endpoints] (abort_source& as) mutable {
+                return with_scheduling_group(sps._task_scheduling_group, [&sps, &as, sync_point_id, target_endpoints = std::move(target_endpoints)] () mutable {
+                    return sps.wait_for_local_hints(as, sync_point_id, target_endpoints);
+                });
+            });
```

Comment:
I don't understand this coordinator/follower thing. Does creating a sync point on the coordinator create sync point on all nodes? What if topology changes during this? What if a node restarts?

Reply:
The coordinator creates follower sync points only on the nodes from the `source_hosts` list. If one of the nodes from this list dies/is removed, the whole operation will fail.